tests/php/ORM/DatabaseTest.php

```diff
@@ -231,4 +237,74 @@ function () use (&$failed) {
         $this->assertInstanceOf('Exception', $ex);
         $this->assertEquals('error', $ex->getMessage());
     }
+
+
+    public function testFieldTypes()
+    {
+        // Scaffold some data
+        $obj = new MyObject();
+        $obj->MyField = "value";
+        $obj->MyInt = 5;
+        $obj->MyFloat = 6.0;
+        $obj->MyBoolean = true;
+        $obj->write();
+
+        $record = DB::prepared_query(
+            'SELECT * FROM "DatabaseTest_MyObject" WHERE "ID" = ?',
+            [ $obj->ID ]
+        )->record();
+
+        // IDs and ints are returned as ints
+        $this->assertInternalType('int', $record['ID']);
+        $this->assertInternalType('int', $record['MyInt']);
+
+        $this->assertInternalType('float', $record['MyFloat']);
+
+        // Booleans are returned as ints – we follow MySQL's lead
+        $this->assertInternalType('int', $record['MyBoolean']);
+
+        // Strings and enums are returned as strings
+        $this->assertInternalType('string', $record['MyField']);
+        $this->assertInternalType('string', $record['ClassName']);
+
+        // Dates are returned as strings
+        $this->assertInternalType('string', $record['Created']);
+        $this->assertInternalType('string', $record['LastEdited']);
+    }
+
+    public function testReadOnlyTransaction()
+    {
+        if (DB::get_conn()->supportsTransactions() == true) {
+            $page = new MyObject();
+            $page->MyField = 'Read only success';
+            $page->write();
+
+            DB::get_conn()->transactionStart('READ ONLY');
+
+            try {
+                $page = new MyObject();
+                $page->MyField = 'Read only page failed';
+                $page->write();
+                DB::get_conn()->transactionEnd();
+
+            } catch (Exception $e) {
+                //could not write this record
+                //We need to do a rollback or a commit otherwise we'll get error messages
+                DB::get_conn()->transactionRollback();
+            }
+
+            DataObject::flush_and_destroy_cache();
+
+            $success = DataObject::get(MyObject::class, "\"MyField\"='Read only success'");
+            $fail = DataObject::get(MyObject::class, "\"MyField\"='Read only page failed'");
+
+            //This page should be in the system
+            $this->assertTrue(is_object($success) && $success->exists());
+
+            //This page should NOT exist, we had 'read only' permissions
+            $this->assertFalse(is_object($fail) && $fail->exists());
```

Comment:
What do you think about
```suggestion
            $this->assertNotInternalType('object', $fail);
            $this->assertFalse($fail->exists()); 
```

Note that second assertion will probably fail because `$fail` isn't an object, so maybe omit it. We could do the opposite assertions for the `assertTrue()` line above